### PR TITLE
Pool, Pool_Stable: adds verification if the current timestamp is after a desired timestamp. 

### PR DESCRIPTION
**Pull request by gangov**
Lints, fixes test and adds a unit tests

### CODE REVIEW

**Comment by gangov** [comment]
lots of changes, I should probably start splitting PR's like this one in smaller chunks. 

Most noticable changes: 
* `packages/phoenix/utils.rs` - new macro that verifies if the current timestamp is within the boundaries set by the user
* `pool` and `pool_stable` contract.rs files - uses the new macro to check that we are within the specified timeframe when interacting with `provide_liquidity`, `withdraw_liquidity` and `swap`
* `factory` and `multihop` - lint
* `tests` - lints and adds test cases for the happy and bad paths when using this macro in `pool` and `pool_stable`

File: contracts/pool/src/tests/swap.rs [hunk 1]
```diff
@@ -1027,8 +1061,198 @@ fn test_should_fail_when_invalid_ask_asset_min_amount() {
...
+                address: share_token_address.clone(),
+                amount: 1_000_000i128,
+            },
+            stake_address: result.clone().stake_address,
+        }
+    );
+    assert_eq!(token1.balance(&user1), 999); // -1 from the swap
+    assert_eq!(token2.balance(&user1), 1001); // 1 from the swap
+
+    // this time 100 units
```
Comment (ueco-jb) [inline contracts/pool/src/tests/swap.rs:1162]: What is this comment referring to?

File: contracts/pool/src/tests/swap.rs [hunk 2]
```diff
@@ -1027,8 +1061,198 @@ fn test_should_fail_when_invalid_ask_asset_min_amount() {
...
+        }
+    );
+    assert_eq!(output_amount, 1000);
+    assert_eq!(token1.balance(&user1), 1999); // 999 + 1_000 as a result of swap
+    assert_eq!(token2.balance(&user1), 1001 - 1000); // user1 sold 1k of token B on second swap
+}
+
+#[test]
+#[should_panic(expected = "Pool: Swap: Transaction executed after deadline!")]
+fn simple_swap_with_should_fail_when_after_the_deadline() {
```
Comment (ueco-jb) [inline contracts/pool/src/tests/swap.rs:1200]: This is fine, but you need  to cover providing and withdrawing liquidity as well.

File: contracts/pool_stable/src/tests/swap.rs [hunk 3]
```diff
@@ -284,3 +325,160 @@ fn swap_simulation_even_pool() {
...
+        factory,
+        None,
+    );
+
+    token1.mint(&user1, &1_001_000);
+    token2.mint(&user1, &1_001_000);
+    env.ledger().with_mut(|li| li.timestamp = 49);
+    pool.provide_liquidity(&user1, &1_000_000, &1_000_000, &None, &Some(50));
+
+    // true means "selling A token"
```
Comment (ueco-jb) [inline contracts/pool_stable/src/tests/swap.rs:364]: Outdated comment.

File: contracts/pool_stable/src/tests/swap.rs [hunk 4]
```diff
@@ -284,3 +325,160 @@ fn swap_simulation_even_pool() {
...
+    // selling just one token with 1% max spread allowed
+    let spread = 100i64; // 1% maximum spread allowed
+                         // making the swap at the final moment
```
Comment (ueco-jb) [inline contracts/pool_stable/src/tests/swap.rs:367]: Those comments also doesn't belong to this  test case. Clear this up.

File: packages/phoenix/src/utils.rs [hunk 5]
```diff
@@ -33,6 +33,11 @@ macro_rules! validate_bps {
     }
 }
 
+/// Validate if the current timestamp is within desired timestamp
+pub fn is_tx_active(env: &Env, deadline: u64) -> bool {
+    env.ledger().timestamp() < deadline
+}
```
Comment (ueco-jb) [inline packages/phoenix/src/utils.rs:39]:
```suggestion
pub fn is_before_deadline(env: &Env, deadline: u64) -> bool {
    env.ledger().timestamp() <= deadline
}
```

Also I wonder if there is a point of creating a function for a one line if statement.

**Review by ueco-jb — COMMENTED** [review]
Minor comments.

File: contracts/pool/src/tests/swap.rs (hunk 1 above)
Comment (gangov) [reply contracts/pool/src/tests/swap.rs]: I believe the test was used with different values. Unnecessary comment has been removed

File: packages/phoenix/src/utils.rs (hunk 5 above)
Comment (gangov) [reply packages/phoenix/src/utils.rs:39]: will get rid of the helper, doesn't do much

**Comment by gangov** [comment]
`env.ledger().timestamp() > deadline` - without `=` should be fine, since we're reversing

**Review by ueco-jb — APPROVED** [review]
LGTM

**Review by ueco-jb — COMMENTED** [review]
Mention this in MIGRATE.md and CHANGELOG and you can merge it.

**Pull request merged**